File: src/duet/DuetAsyncFIFO.cc
#include "duet/DuetAsyncFIFO.hh"
#include "duet/DuetAsyncFIFOCtrl.hh"

namespace gem5 {
namespace duet {

void DuetAsyncFIFO::UpstreamPort::recvFunctional ( PacketPtr pkt ) {
    _owner->_downstream_port.sendFunctional ( pkt );
}

bool DuetAsyncFIFO::UpstreamPort::recvTimingReq ( PacketPtr pkt ) {
    return _owner->_upstream_ctrl->recv ( pkt );
}

void DuetAsyncFIFO::UpstreamPort::recvRespRetry () {
    _owner->_upstream_ctrl->try_send ();
}

bool DuetAsyncFIFO::UpstreamPort::recvTimingSnoopResp ( PacketPtr pkt ) {
    return _owner->_upstream_ctrl->recv_snoop ( pkt );
}

bool DuetAsyncFIFO::DownstreamPort::recvTimingResp ( PacketPtr pkt ) {
    return _owner->_downstream_ctrl->recv ( pkt );
}

void DuetAsyncFIFO::DownstreamPort::recvReqRetry () {
    _owner->_downstream_ctrl->try_send ();
}

void DuetAsyncFIFO::DownstreamPort::recvRetrySnoopResp () {
    _owner->_downstream_ctrl->try_send ( true );
}

void DuetAsyncFIFO::DownstreamPort::recvTimingSnoopReq ( PacketPtr pkt ) {
    _owner->_downstream_ctrl->recv_snoop ( pkt );
}

DuetAsyncFIFO::DuetAsyncFIFO ( const DuetAsyncFIFOParams & p )
    : SimObject             ( p )
    , _stage                ( p.stage )
    , _capacity             ( p.capacity )
    , _is_snooping          ( p.snooping )
    , _upstream_port        ( p.name + ".upstream_port", this )
    , _downstream_port      ( p.name + ".downstream_port", this )
    , _upstream_ctrl        ( p.upstream_ctrl )
    , _downstream_ctrl      ( p.downstream_ctrl )
    , _downward_fifo        ()
    , _upward_fifo          ()
{
    _upstream_ctrl->_owner          = this;
    _upstream_ctrl->_is_snooping    = p.snooping;
    _upstream_ctrl->_push           = p.capacity;

    _downstream_ctrl->_owner        = this;
    _downstream_ctrl->_is_snooping  = p.snooping;
    _downstream_ctrl->_push         = p.capacity;
}

Port & DuetAsyncFIFO::getPort ( const std::string & if_name
        , PortID idx ) {
    panic_if ( idx != InvalidPortID,
            "DuetAsyncFIFO does not support vector ports." );

    if ( if_name == "upstream_port" )
        return _upstream_port;
    else if ( if_name == "downstream_port" )
        return _downstream_port;
    else
        return SimObject::getPort ( if_name, idx );
}

}   // namespace duet
}   // namespace gem5

File: src/duet/DuetAsyncFIFO.py
from m5.params import *
from m5.proxy import Parent
from m5.SimObject import SimObject
from m5.objects.ClockedObject import ClockedObject

class DuetAsyncFIFOCtrl (ClockedObject):
    type            = "DuetAsyncFIFOCtrl"
    cxx_class       = "gem5::duet::DuetAsyncFIFOCtrl"
    cxx_header      = "duet/DuetAsyncFIFOCtrl.hh"

    is_upstream     = Param.Bool ("If this is an upstream ctrl or a downstream ctrl")

class DuetAsyncFIFO (SimObject):
    type            = "DuetAsyncFIFO"
    cxx_class       = "gem5::duet::DuetAsyncFIFO"
    cxx_header      = "duet/DuetAsyncFIFO.hh"

    stage           = Param.Cycles      ( 2, "Synchronizer stage" )
    capacity        = Param.Unsigned    ( 64, "FIFO capacity (per direction)" )
    upstream_port   = ResponsePort      ( "Response port to upstream" )
    downstream_port = RequestPort       ( "Request port to downstream" )
    upstream_ctrl   = Param.DuetAsyncFIFOCtrl ( "Upstream ctrl" )
    downstream_ctrl = Param.DuetAsyncFIFOCtrl ( "Downstream ctrl" )
    snooping        = Param.Bool        ( False, "If the Async FIFO supports snooping across" )

    def __init__ (self, **kwargs):
        upstream_clk_domain = kwargs.pop ("upstream_clk_domain", None)
        downstream_clk_domain = kwargs.pop ("downstream_clk_domain", None)

        super().__init__(**kwargs)

        self.upstream_ctrl = DuetAsyncFIFOCtrl ( is_upstream = True )
        if upstream_clk_domain:
            self.upstream_ctrl.clk_domain = upstream_clk_domain

        self.downstream_ctrl = DuetAsyncFIFOCtrl ( is_upstream = False )
        if downstream_clk_domain:
            self.downstream_ctrl.clk_domain = downstream_clk_domain

    def __getattr__ (self, attr):
        if attr == "upstream_clk_domain":
            return self.upstream_ctrl.clk_domain
        elif attr == "downstream_clk_domain":
            return self.downstream_ctrl.clk_domain
        else:
            return super().__getattr__(attr)

    def __setattr__ (self, attr, val):
        if attr == "upstream_clk_domain":
            self.upstream_ctrl.clk_domain = val
        elif attr == "downstream_clk_domain":
            self.downstream_ctrl.clk_domain = val
        else:
            super().__setattr__(attr, val)
